src/server/src/main/java/io/cassandrareaper/storage/cassandra/CassandraStorage.java

```diff
@@ -0,0 +1,836 @@
+  private PreparedStatement deleteHeartbeatPrepStmt;
+
+  private final RepairUnitDao repairUnitDao;
+  private final RepairScheduleDao repairScheduleDao;
+  private final ClusterDao clusterDao;
+  private final EventsDao eventsDao;
+  private final MetricsDao metricsDao;
+  private final Concurrency concurrency;
+  private final SnapshotDao snapshotDao;
+  private final OperationsDao operationsDao;
+
+  public CassandraStorage(
+      UUID reaperInstanceId,
+      ReaperApplicationConfiguration config,
+      Environment environment,
+      CassandraMode mode) throws ReaperException {
+
+    this.reaperInstanceId = reaperInstanceId;
+    this.defaultTimeout = config.getHangingRepairTimeoutMins();
+    CassandraFactory cassandraFactory = config.getCassandraFactory();
+    overrideQueryOptions(cassandraFactory, mode);
+    overrideRetryPolicy(cassandraFactory);
+    overridePoolingOptions(cassandraFactory);
+
+    // https://docs.datastax.com/en/developer/java-driver/3.5/manual/metrics/#metrics-4-compatibility
+    cassandraFactory.setJmxEnabled(false);
+    if (!CassandraStorage.UNINITIALISED.compareAndSet(true, false)) {
+      // If there's been a past connection attempt, metrics are already registered
+      cassandraFactory.setMetricsEnabled(false);
+    }
+
+    cassandra = cassandraFactory.build(environment);
+    if (config.getActivateQueryLogger()) {
+      cassandra.register(QueryLogger.builder().build());
+    }
+    CodecRegistry codecRegistry = cassandra.getConfiguration().getCodecRegistry();
+    codecRegistry.register(new DateTimeCodec());
+    session = cassandra.connect(config.getCassandraFactory().getKeyspace());
+    version = cassandra.getMetadata().getAllHosts()
+        .stream()
+        .map(h -> h.getCassandraVersion())
+        .min(VersionNumber::compareTo)
+        .get();
+
+    this.eventsDao =  new EventsDao(session);
+    this.metricsDao  = new MetricsDao(session);
+    this.snapshotDao = new SnapshotDao(session);
+    this.operationsDao  = new OperationsDao(session);
+    this.concurrency = new Concurrency(version, reaperInstanceId, session);
+    this.repairUnitDao  = new RepairUnitDao(defaultTimeout, session);
+    this.repairSegmentDao = new RepairSegmentDao(concurrency, repairUnitDao, session);
+    this.repairScheduleDao = new RepairScheduleDao(repairUnitDao, session);
+    this.clusterDao  = new ClusterDao(repairScheduleDao, repairUnitDao, eventsDao, session, objectMapper);
+    this.repairRunDao =  new RepairRunDao(repairUnitDao, clusterDao, repairSegmentDao, session);
+
+    boolean skipMigration = System.getenv().containsKey("REAPER_SKIP_SCHEMA_MIGRATION")
+        ? Boolean.parseBoolean(System.getenv("REAPER_SKIP_SCHEMA_MIGRATION"))
+        : Boolean.FALSE;
+
+    if (skipMigration) {
+      LOG.info("Skipping schema migration as requested.");
+    } else {
+      MigrationManager.initializeAndUpgradeSchema(cassandra, session, config, version, mode);
+    }
```

Comment:
```suggestion
    version = cassandra.getMetadata().getAllHosts()
        .stream()
        .map(h -> h.getCassandraVersion())
        .min(VersionNumber::compareTo)
        .get();

    boolean skipMigration = System.getenv().containsKey("REAPER_SKIP_SCHEMA_MIGRATION")
        ? Boolean.parseBoolean(System.getenv("REAPER_SKIP_SCHEMA_MIGRATION"))
        : Boolean.FALSE;

    if (skipMigration) {
      LOG.info("Skipping schema migration as requested.");
    } else {
      MigrationManager.initializeAndUpgradeSchema(cassandra, session, config, version, mode);
    }

    this.eventsDao =  new EventsDao(session);
    this.metricsDao  = new MetricsDao(session);
    this.snapshotDao = new SnapshotDao(session);
    this.operationsDao  = new OperationsDao(session);
    this.concurrency = new Concurrency(version, reaperInstanceId, session);
    this.repairUnitDao  = new RepairUnitDao(defaultTimeout, session);
    this.repairSegmentDao = new RepairSegmentDao(concurrency, repairUnitDao, session);
    this.repairScheduleDao = new RepairScheduleDao(repairUnitDao, session);
    this.clusterDao  = new ClusterDao(repairScheduleDao, repairUnitDao, eventsDao, session, objectMapper);
    this.repairRunDao =  new RepairRunDao(repairUnitDao, clusterDao, repairSegmentDao, session);
```

Schema migrations need to happen before creating the DAOs. It looks like their constructors invoke `prepareStatements()` which will do exactly that but on non existing tables.

After moving this block I was able to start Reaper properly.